### ISSUE TITLE
Add "kubelet cluster update"

### ISSUE THREAD

**Issue by maratoid** [samsung-cnct/k2#85]
Kubelet cluster update should be a new ansible role that is not part of the default set and must be called explicitly and has a selector based on provider.  The ansible role should perform the upgrade/downgrade (should be the same process) tasks as laid out in the following tickets:
- gke: https://github.com/samsung-cnct/k2/issues/217#issuecomment-285467380 
- aws: https://github.com/samsung-cnct/k2/issues/218 (see last comment for additional detail.  if there is confusion, please speak to @coffeepac)
-- this may be completed best by completing #60 first

**Comment by coffeepac** [comment]
this is blocked by #218 #219

**Comment by coffeepac** [comment]
use the information from here:  https://github.com/samsung-cnct/k2/issues/217#issuecomment-285467380

**Comment by guineveresaenger** [comment]
@maratoid @coffeepac @DStorck 

I've been trying to get my head wrapped into this and have a few questions.
- "kubelet cluster update" refers to what, exactly? I'm confused about the use of the work kubelet here.
- In K2, we seem to be handling AWS and GKE entirely differently: GKE update lives in kraken.provider.gke role as a regular Task, whereas we have an entire role called kraken.upgrade that holds an .aws subfolder (but nothing pertaining to gke). This makes sense given the different ways both providers are structured; I just wanted to see if there were any more thoughts and comments on this.
- We have both update.yaml and upgrade.yaml. What is the difference? Also, there seems to be an unresolved merge conflict in upgrade.yaml.

**Comment by guineveresaenger** [comment]
recap from chat: the work for master nodes is done; I am researching the work for non-master and non-etc nodes.

**Comment by guineveresaenger** [comment]
Starting with aws, as it is being handled differently:
- in kraken.upgrade.aws, there's a task for deleting control plane nodes.
- there is a kill node filter plugin that does this. Because it was written with master nodes in mind, it has a builtin wait period.
- it would seem DRY to use that same filter on non control plane nodes; however this would add extra wait time that I'm not sure we need in this case (cluster nodes can all die and be respawned without the cluster breaking)
- it looks like we want a task for updating non control plane nodes that executes in main after the control plain node tasks have run.
Question:
- Do we want to have a "wait to respawn" for cluster nodes? While killing all at once won't break the cluster, could this result in pods breaking or being lost? When I ran this manually a couple months ago, this didn't happen, but they were pretty small, empty pods. Is this a concern?
@coffeepac just tagging you for reminder purposes

**Comment by coffeepac** [comment]
We do want there to be a wait so that workloads can migrate smoothly.  Terminating all nodes in a cluster and respawning them is fast but that's inappropriate behavior.  The cluster would still work but while the nodes respawned any apps that should be running on the cluster wouldn't be.

**Comment by guineveresaenger** [comment]
Sanity check: We have five node pools. We want to update master(already done), clusterNodes, and specialNodes. Yes? No on both etcd and etcdEvents, correct?

**Comment by coffeepac** [comment]
@guineveresaenger correct.  No to both etcd and etcdEvents

**Comment by guineveresaenger** [comment]
GKE upgrade will happen directly on `/up.sh` with a new k8s version in the config file; no further action needed. Since GKE has its own control plane, there are also some restrictions on which k8s versions the nodes can be - please consult the most recent [release notes](https://cloud.google.com/container-engine/release-notes) for this.

Documentation for `upgrade` should include noting that this is an AWS action only.

**Comment by coffeepac** [comment]
@guineveresaenger if you manage to get all of the AWS upgrade to function from upgrade.sh then we'll want to have it function for GKE as well.

in the nearish future, we'll be restricting what can be updated when running up.sh as its rather dangerous right now.

**Comment by guineveresaenger** [comment]
It's functioning right now; I'm just adding some false input checks. To disallow passing in bogus node names and printing a fail message when users do so.

Currently, the GKE upgrade happens automatically on up with a gke cluster. I take it we do not want this; this means separating it out into its own role, correct?

**Comment by coffeepac** [comment]
woo!

what happens if you call upgrade with a GKE cluster right now?

**Comment by guineveresaenger** [comment]
Sorry; out for lunch
If you look at upgrade.yaml, it only calls kraken.config, nodepoolSelecor, and also upgrade.aws roles. nodepoolSelector doesn't do anything right now, and kraken.config probably handles setting k8s minor versions for use in upgrade.aws. So, upgrade doesn't really do anything with GKE specifically. 

Upgrading GKE clusters happens in the same place where we create the cluster, so running upgrade is redundant and won't do anything. (The latter is a guess; I brought down my gke cluster before I could try it out).

**Comment by guineveresaenger** [comment]
Digging deeper into how GKE is handled.
GKE updates get handled in `update-cluster.yaml` (in kraken.provider.gke) upon changes in the config file. 
There are 2 tasks in `update-cluster.yaml` - one for updating master, and one for nodes.
### Updating nodes:
Each GKE nodepool has its own kubeconfig in the config file. The way to upgrade k8s version by nodepool is to assign a different version to each nodepool's kubeconfig in the config file, and run `./up.sh` or  `./upgrade.sh`. This works; it works well, although it does not make for user interaction beyond messing with the config file. The difficulty is that updating it this way currently throws the following error on each node:

`"ERROR: (gcloud.container.clusters.upgrade) ResponseError: code=400, message=Operation operation-1496877190764-93123b55 is currently upgrading cluster guinnode. Please wait and try again once it's done."`

and then exits the playbook.

It does exactly what it is supposed to do - update all the nodes, carefully, one by one - but it doesn't really give a good message back to the user. Ideas would be appreciated.

### Updating master
I am confused by this. The `Update master, if needed` task calls upgrade - on a cluster, with `--master` as the flag. The conditions on the task, however, are such that currently it is impossible to update the control plane, since the current gke config file does not supply a kubeConfig in providerConfig stanza. Since IIRC, there are some limits imposed by GKE on which versions are supported for upgrading master at any given time, it makes sense to me to not allow for updates of master. But this task seems extraneous. I see we can also update other things with this task (like zone changes) but currently all those options are blocked by the when statement demanding `cluster.providerConfig.kubeConfig.version` to be defined.

**Comment by guineveresaenger** [comment]
I am wondering if I should restructure k2 to update GKE clusters in the kraken.update task. It seems more in parallele to what happens with aws. To clear up the above issue where only one nodepool gets updated, the task in question should include a wait period to check whether the current nodepool has been updated, so that the process doesn't exit early. I'm thinking as of right now, this should be part of the shell command? I'm imagining this:

1. Update first nodepool
2. Wait for all nodepools of that name to have been updated by checking the output to kubectl command and comparing their versions (which are thankfully part of the output)
3. when all nodes in pool have upgraded, continue.

Considerations: This practically begs for using a nodepool flag, just like AWS has. Instead of upgrading all nodepools, we can pull in nodepools from flag. 

@patthec since this is a considerable change to k2, does this sound like a valid approach?

**Comment by coffeepac** [comment]
A bit late, but to close the documented loop:  yes, please move the gke update task from the gke provider to role to the update role.  

For GKE do not attempt to update the control plane.  We as users of GKE have no control over what the control plane is running.  GKE controls that knob entirely.

**Comment by guineveresaenger** [comment]
I just made some astonishing discoveries.
So...it turns out that update-cluster exists in create-cluster, because that is _the way_ to get a version on the cluster that we want; otherwise we will get the current gke default version. (Aside: this is why running up a gke cluster takes so long!) I'm currently testing if it works for the control plane (as advertised), but it doesn't look like that task ever executes. Here is the task, for reference:
`- name: Update master, if needed
  command: >
    gcloud container clusters upgrade {{ cluster.name }}
      --master
      --project {{ cluster.providerConfig.project }}
      --zone {{ cluster.providerConfig.zone.primaryZone }}
      --no-async
      --quiet
  when: (cluster.providerConfig.kubeConfig is defined) and (cluster.providerConfig.kubeConfig.version is defined)`

Unless there's something about this I don't know about, I would carefully suggest deleting that task. I do think that it is possible to update the control plane somewhat (there's documentation on this at gce) so maybe it's not completely useless? but in the current gke config file, we do not set kubeConfig in providerConfig stanza.

**Comment by guineveresaenger** [comment]
update: yeah, the master remains untouched.

**Comment by guineveresaenger** [comment]
PR #44 ready for review:
- check if clusters can be upgraded on both AWS and GKE
- check if documentation in README describing how to upgrade makes sense
